src/metabase/lib/fe_util.cljc

```diff
@@ -71,54 +87,177 @@
+    segment
+    {:lib/type :metadata/segment
+     :id (last segment-ref)
+     :display-name (i18n/tru "Unknown Segment")}))
+
+(defn- metric-metadata-from-ref
+  [query metric-ref]
+  (if-let [metric (lib.metadata/metric query (last metric-ref))]
+    metric
+    {:lib/type :metadata/metric
+     :id (last metric-ref)
+     :display-name (i18n/tru "Unknown Metric")}))
+
+(defmulti expression-parts-method
+  "Builds the expression parts by dispatching on the type of the argument."
+  {:arglists '([query stage-number arg])}
+  (fn [_query _stage-number value]
+    (lib.dispatch/dispatch-value value))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod expression-parts-method :default
+  [query stage-number value]
+  (let [[op options & args] (cond-> value
+                              (expandable-temporal-expression? value) expand-temporal-expression
+                              (case-or-if-expression? value) flatten-case-or-if-expression)]
+    {:lib/type :mbql/expression-parts
+     :operator op
+     :options  options
+     :args     (mapv #(expression-parts-method query stage-number %) args)}))
+
+(defmethod expression-parts-method :dispatch-type/expression-parts
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/string
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/integer
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/number
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/boolean
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :dispatch-type/keyword
+  [_query _stage-number value]
+  (name value))
+
+(defmethod expression-parts-method :dispatch-type/nil
+  [_query _stage-number value]
+  value)
+
+(defmethod expression-parts-method :field
+  [query stage-number field-ref]
+  (column-metadata-from-ref query stage-number field-ref))
+
+(defmethod expression-parts-method :metadata/column
+  [_query _stage-number column]
+  column)
+
+(defmethod expression-parts-method :segment
+  [query _stage-number segment-ref]
+  (segment-metadata-from-ref query segment-ref))
+
+(defmethod expression-parts-method :metadata/segment
+  [_query _stage-number segment]
+  segment)
+
+(defmethod expression-parts-method :metric
+  [query _stage-number metric-ref]
+  (metric-metadata-from-ref query metric-ref))
+
+(defmethod expression-parts-method :metadata/metric
+  [_query _stage-number metric]
+  metric)
+
+(defmethod expression-parts-method :expression
+  [query stage-number expression-ref]
+  (column-metadata-from-ref query stage-number expression-ref))
+
+(mu/defn expression-parts :- [:or ExpressionArg ExpressionParts]
+  "Return the parts of the filter clause `arg` in query `query` at stage `stage-number`."
+  ([query value]
+   (expression-parts-method query -1 value))
```

Comment:
Typically these other arities would call the top-level function, rather than the method directly, in case there's any pre/post-processing: `(expression-parts query -1 value)`.